src/FEValues/interface_values.jl

```diff
@@ -0,0 +1,265 @@
+
+`normal_dotted::Bool` determines whether to use the definition ``\\llbracket v \\rrbracket=v^- -v^+`` if it's `false`, or
+ the definition  ``\\llbracket v \\rrbracket=v^- ⋅ \\vec{n}^- + v^+ ⋅ \\vec{n}^+`` if it's `true`, which is the default.
+
+!!! note
+    If `normal_dotted == true` then the jump of scalar shape values is a vector.
+"""
+shape_value_jump
+
+"""
+    shape_gradient_average(iv::InterfaceValues, qp::Int, base_function::Int)
+
+Compute the average of the shape function gradient at the quadrature point over the interface.
+"""
+shape_gradient_average
+
+"""
+    shape_gradient_jump(iv::InterfaceValues, qp::Int, base_function::Int, normal_dotted::Bool = true)
+
+Compute the jump of the shape function gradient at the quadrature point over the interface.
+
+`normal_dotted::Bool` determines whether to use the definition ``\\llbracket \\vec{v} \\rrbracket=\\vec{v}^- -\\vec{v}^+`` if it's `false`, or
+ the definition  ``\\llbracket \\vec{v} \\rrbracket=\\vec{v}^- ⋅ \\vec{n}^- + \\vec{v}^+ ⋅ \\vec{n}^+`` if it's `true`, which is the default.
+
+!!! note
+    If `normal_dotted == true` then the jump of the gradient vector is a scalar.
+"""
+shape_gradient_jump
+
+for (func,                      f_,                 multiplier  ) in (
+    (:shape_value,              :shape_value,       :(1),       ),
+    (:shape_value_average,      :shape_value,       :(0.5),     ),
+    (:shape_gradient,           :shape_gradient,    :(1),       ),
+    (:shape_gradient_average,   :shape_gradient,    :(0.5),     ),
+)
+    @eval begin
+        function $(func)(iv::InterfaceValues, qp::Int, i::Int)
+            nbf = getnbasefunctions(iv)
+            if i <= nbf/2
+                fv = iv.face_values
+                f_value = $(f_)(fv, qp, i)
+                return $(multiplier) * f_value
+            elseif i <= nbf
+                fv = iv.face_values_neighbor
+                f_value = $(f_)(fv, qp, i - nbf ÷ 2)
+                return $(multiplier) * f_value
+            end
+            error("Invalid base function $i. Interface has only $(nbf) base functions")
+        end
+    end
+end
+
+for (func,                      f_,             ) in (
+    (:shape_value_jump,         :shape_value,   ),
+    (:shape_gradient_jump,      :shape_gradient,),
+)
+    @eval begin
+        function $(func)(iv::InterfaceValues, qp::Int, i::Int, normal_dotted::Bool = true)
+            f_value = $(f_)(iv, qp, i)
+            nbf = getnbasefunctions(iv)
+            if i <= nbf/2
+                normal_dotted || return f_value
+                multiplier = getnormal(iv, qp, true)
+                return f_value isa Number ? f_value * multiplier : f_value ⋅ multiplier
+            else
+                normal_dotted || return -f_value
+                multiplier = getnormal(iv, qp, false)
+                return f_value isa Number ? f_value * multiplier : f_value ⋅ multiplier
+            end
+        end
+    end
+end
```

Comment:
As @fredrikekre and @termi-official  pointed out, [some use the definition with normals ](https://www.researchgate.net/publication/211746421_Unified_Analysis_of_Discontinuous_Galerkin_Methods_for_Elliptic_Problems)and [some use (here - there) definition](https://onlinelibrary.wiley.com/doi/pdf/10.1002/nme.1667). This implements both and defaults to using normals.

Reply:
While I fully understand the rationale I am not sure what the performance implications are, because the function now has different return values depending on the input parameters. Fredrik might can give you more information on this.

Reply:
If this is a problem, then lets use the "problematic" definition without normal dotting and leave the responsibility to the user to do the right thing for their formulation.

Reply:
Doing simple testing with `@Benchmark`:
```julia
julia> @benchmark shape_value_jump(iv, 1, 1, true)
BenchmarkTools.Trial: 10000 samples with 600 evaluations.
 Range (min … max):  199.667 ns …  3.141 μs  ┊ GC (min … max): 0.00% … 91.83%
 Time  (median):     207.167 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   214.168 ns ± 94.643 ns  ┊ GC (mean ± σ):  1.47% ±  3.16%

  ▅█▂▁▄▅▇▂▂▄▂          ▁▂▂▁ ▁▁▁▁                               ▂
  ████████████▇██▇▅▇▄▆██████████▇▆▆▅▅▅▆▄▆▆▆▆▅▅▁▅▆▆▆▅▅▁▆▅▄▃▅▅▃▅ █
  200 ns        Histogram: log(frequency) by time       303 ns <

 Memory estimate: 96 bytes, allocs estimate: 4.
 
 julia> @benchmark shape_value_jump(iv, 1, 1, false)
BenchmarkTools.Trial: 10000 samples with 870 evaluations.
 Range (min … max):  136.092 ns …  1.361 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     142.874 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   146.104 ns ± 30.703 ns  ┊ GC (mean ± σ):  0.39% ± 2.07%

  ▃█▆▄▂▇▆▄▁▁ ▁▂▂▃▄▂▁▁▂▃                                        ▂
  ██████████████████████▇▇▇▇▅▆▆▅▆▃▅▃▅▄▄▄▄▄▅▄▄▄▃▃▄▄▃▄▄▄▄▃▃▄▄▄▄▃ █
  136 ns        Histogram: log(frequency) by time       215 ns <

 Memory estimate: 32 bytes, allocs estimate: 2.
```

Which isn't far from that of `FaceVakues` (in terms of allocation, 2~3 times slower in terms of time? I mean it does operations not just accessing tensor)?

```julia
julia> @benchmark shape_value(iv.face_values_a, 1, 1)
BenchmarkTools.Trial: 10000 samples with 980 evaluations.
 Range (min … max):  59.184 ns …   3.115 μs  ┊ GC (min … max): 0.00% … 97.14%
 Time  (median):     62.245 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   74.055 ns ± 104.818 ns  ┊ GC (mean ± σ):  5.41% ±  3.97%

  █▇▆▅▁     ▁▂                                                 ▁
  ██████▇▇▆▆████▆▅▆▅▄▅▅▅▄▄▃▄▆▅▅▂▅▃▄▄▄▄▄▃▃▃▄▂▄▄▂▃▂▃▄▅▆▇▅▇▇▆▄▄▅▆ █
  59.2 ns       Histogram: log(frequency) by time       231 ns <

 Memory estimate: 80 bytes, allocs estimate: 2.
```

Reply:
updated benchmarks:

```julia
julia> @benchmark shape_value(facevalues1, 1, 1)
BenchmarkTools.Trial: 10000 samples with 995 evaluations.
 Range (min … max):  26.734 ns …  1.063 μs  ┊ GC (min … max): 0.00% … 95.63%
 Time  (median):     27.236 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   29.875 ns ± 18.182 ns  ┊ GC (mean ± σ):  1.00% ±  1.65%

  ▇█▇▄              ▂▁    ▄▄▃  ▃▆▅▄                     ▂▂    ▂
  ████▇▆▄▁▁▁▆▅▆▆█▇▄▄██▅▃▄▇███▇▆██████▇▇▆▇▇▆▆▅▆▆▆▅▆▇▆▆▆▇████▇▅ █
  26.7 ns      Histogram: log(frequency) by time      39.5 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

julia> @benchmark shape_value_average(iv, 1, 1)
BenchmarkTools.Trial: 10000 samples with 993 evaluations.
 Range (min … max):  30.312 ns …  1.178 μs  ┊ GC (min … max): 0.00% … 95.08%
 Time  (median):     30.715 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   33.991 ns ± 19.996 ns  ┊ GC (mean ± σ):  0.92% ±  1.60%

  █▅         ▄▄▁▆▃         ▂▂                                 ▂
  ███▇▁▄▃▁███████████▇▇▇▆█▇██▇▆▅▅▇▆▆▆▆▅▆▅▄▄▄▄▄▅▅▆▄▄▅▄▄▃▃▄▅▁▁▃ █
  30.3 ns      Histogram: log(frequency) by time      57.4 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
 
julia> @benchmark shape_value_jump(iv, 1, 1, true)
BenchmarkTools.Trial: 10000 samples with 993 evaluations.
 Range (min … max):  35.347 ns …  4.546 μs  ┊ GC (min … max): 0.00% … 95.60%
 Time  (median):     38.167 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   62.274 ns ± 88.570 ns  ┊ GC (mean ± σ):  2.21% ±  2.36%

  █▅▃▄▄▂                      ▁▁▁▁▁                           ▁
  ███████▆▆▆▆▆▆▅▅▆▅▆▅▆▆▅▆▆▅▇▇█████████▇███▇▇▇▇█▇▆▇▆▅▆▅▅▅▅▃▄▄▄ █
  35.3 ns      Histogram: log(frequency) by time       253 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.

julia> @benchmark shape_value_jump(iv, 1, 1, false)
BenchmarkTools.Trial: 10000 samples with 993 evaluations.
 Range (min … max):  34.038 ns …  1.072 μs  ┊ GC (min … max): 0.00% … 94.32%
 Time  (median):     35.146 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   38.028 ns ± 18.710 ns  ┊ GC (mean ± σ):  0.79% ±  1.64%

  ▇██▅▄▂    ▁▂▅▆▄▄▄▄▅▃▁   ▂▁▁▁ ▁▁▁                            ▂
  ██████▅▆▅▇████████████▇▇████████▇▄▇▁▆▅▄▄▅▃▄▄▅▅▆▅▄▁▆▅▅▄▄▃▅▅▆ █
  34 ns        Histogram: log(frequency) by time        60 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

 ```